File: DGE/2_DGE_inspection_plots_from_csv.r
# ===============================================================================
# CLUSTER DGE INSPECTION PLOTS SCRIPT
# ===============================================================================
# Purpose:
#   This script generates inspection plots for merged differential gene expression
#   (DGE) results on clusters, cell subtypes, and optionally combined cell types.
#   It is intended to be run after generating merged DGE CSV files.
#
# What are "Inspection Plots"?
#   In this context, "inspection plots" are a set of diagnostic and summary visualizations
#   designed to help users quickly assess, compare, and interpret the results of DGE analyses.
#   These plots provide visual summaries of gene expression changes across different groups
#   or comparisons, highlight the most significant genes, and reveal patterns or outliers
#   that may warrant further investigation.
#
#   Specifically, this script produces:
#     - Scatterplots: Compare log2 fold changes for shared genes between all pairs of comparisons,
#       allowing visual assessment of concordance or divergence in gene regulation.
#     - Stacked Bar Charts: Show the number of up- and down-regulated genes per comparison,
#       summarizing the overall direction and magnitude of differential expression.
#     - Volcano Plots: Display the relationship between fold change and statistical significance
#       for each comparison, highlighting the most strongly regulated genes.
#
# Usage:
#   - Source this script in R after running the DGE analysis and merging steps.
#   - The script expects merged DGE CSV files to be present in the output directories.
#   - Plots and intermediate data will be saved in subdirectories of the output folders.
#
# Main Features:
#   1. Standardizes DGE data columns from different formats.
#   2. Generates scatterplots comparing log2 fold changes for all pairs of comparisons.
#   3. Produces stacked bar charts showing the number of up/down-regulated genes per comparison.
#   4. Creates volcano plots for each comparison, highlighting top up- and down-regulated genes.
#   5. Handles DGE results for clusters, cell types, and optionally combined cell types.
#
# Output:
#   - PDF plots and CSV data files are saved in 'inspection_plots_pdf' and
#     'inspection_plot_data' directories under each DGE output folder.
#
# Dependencies:
#   - Requires ggplot2, ggrepel, dplyr, gtools, and RColorBrewer packages.
#   - Assumes 'util_headers.r' is available for loading project-specific variables.
#
# Author: [Your Name or Lab]
# Date: [YYYY-MM-DD]
# ===============================================================================

# Inspection plots for merged DGE on clusters
# Usage: source this script after generating merged_dge_on_clusters.csv
source("util_headers.r")

library(ggplot2)
library(RColorBrewer)

cluster_dge_inspect_env <- new.env()

# Standardize column names for different DGE file formats
cluster_dge_inspect_env$standardize_dge_columns <- function(dge) {
    # Check if this is the cellTypes format
    if ("names" %in% colnames(dge) && "logfoldchanges" %in% colnames(dge)) {
        # Map cellTypes format to standard format
        dge <- dge |>
            dplyr::rename(
                gene = names,
                avg_log2FC = logfoldchanges,
                p_val_adj = pvals_adj,
                p_val = pvals
            )
        # Add missing columns with default values if they don't exist
        if (!"pct.1" %in% colnames(dge)) dge$pct.1 <- NA
        if (!"pct.2" %in% colnames(dge)) dge$pct.2 <- NA
        # Rename cell_type to region for consistency (if region doesn't exist)
        if ("cell_type" %in% colnames(dge) && !"region" %in% colnames(dge)) {
            dge <- dge |> dplyr::rename(region = cell_type)
        }
    }

    # Ensure avg_log2FC is numeric
    dge$avg_log2FC <- as.numeric(dge$avg_log2FC)
    dge
}


# Load merged DGE data
cluster_dge_inspect_env$load_dge_data <- function(dge_csv_path) {
    dge <- read.csv(dge_csv_path, stringsAsFactors = FALSE)
    # Standardize column names based on format
    dge <- cluster_dge_inspect_env$standardize_dge_columns(dge)
    dge
}


# 1. Scatterplots for all pairs of comparisons (shared genes)
cluster_dge_inspect_env$plot_comparison_scatterplots <- function(dge_subset, lfc_threshold = 0.25, pval_threshold = 0.05) {
    comparisons <- unique(dge_subset$comparison)
    combn_pairs <- gtools::combinations(length(comparisons), 2, comparisons)
    for (i in seq_len(nrow(combn_pairs))) {
        comp1 <- combn_pairs[i, 1]
        comp2 <- combn_pairs[i, 2]
        d1 <- dge_subset |> dplyr::filter(comparison == comp1, p_val_adj < pval_threshold)
        d2 <- dge_subset |> dplyr::filter(comparison == comp2, p_val_adj < pval_threshold)
        shared_genes <- intersect(d1$gene, d2$gene)
        if (length(shared_genes) == 0) next

        d1 <- d1 |>
            dplyr::filter(gene %in% shared_genes) |>
            dplyr::select(gene, avg_log2FC) |>
            dplyr::rename(avg_log2FC_1 = avg_log2FC) |>
            dplyr::distinct(gene, .keep_all = TRUE)

        d2 <- d2 |>
            dplyr::filter(gene %in% shared_genes) |>
            dplyr::select(gene, avg_log2FC) |>
            dplyr::rename(avg_log2FC_2 = avg_log2FC) |>
            dplyr::distinct(gene, .keep_all = TRUE)

        plot_df <- dplyr::left_join(d1, d2, by = "gene")

        plot_df$size <- abs(plot_df$avg_log2FC_1) + abs(plot_df$avg_log2FC_2)
        write.csv(plot_df, file = file.path(cluster_dge_inspect_env$plot_data_dir, paste0("scatter_", comp1, "_against_", comp2, ".csv")), row.names = FALSE)
        p <- ggplot(plot_df, aes(x = avg_log2FC_1, y = avg_log2FC_2, label = gene)) +
            ggplot2::geom_hline(yintercept = 0, linetype = "dashed", color = "gray50", alpha = 0.7) +
            ggplot2::geom_vline(xintercept = 0, linetype = "dashed", color = "gray50", alpha = 0.7) +
            ggplot2::geom_point(aes(size = size), alpha = 0.6, color = "darkred") +
            ggrepel::geom_text_repel(max.overlaps = 10, size = 2) +
            ggplot2::labs(
                title = paste("Shared DE Genes:", comp1, "vs", comp2),
                x = paste0("avg_log2FC (", comp1, ")"),
                y = paste0("avg_log2FC (", comp2, ")")
            ) +
            ggplot2::theme_minimal()
        ggplot2::ggsave(file.path(cluster_dge_inspect_env$plot_pdf_dir, paste0("scatter_", comp1, "_against_", comp2, ".pdf")), p, width = 7, height = 7)
    }
}

# 2. Stacked bar chart: number of up/down-regulated genes per comparison
cluster_dge_inspect_env$plot_stacked_barchart <- function(dge_subset, lfc_threshold = 0.25, pval_threshold = 0.05) {
    dge_subset <- dge_subset |>
        dplyr::mutate(regulation = dplyr::case_when(
            avg_log2FC >= lfc_threshold & p_val_adj < pval_threshold ~ "Up",
            avg_log2FC <= -lfc_threshold & p_val_adj < pval_threshold ~ "Down",
            TRUE ~ "NS"
        ))
    bar_df <- dge_subset |>
        dplyr::filter(regulation != "NS") |>
        dplyr::group_by(comparison, regulation) |>
        dplyr::summarise(n_genes = dplyr::n(), .groups = "drop")
    write.csv(bar_df, file = file.path(cluster_dge_inspect_env$plot_data_dir, "stacked_barchart_data.csv"), row.names = FALSE)
    p <- ggplot(bar_df, aes(x = comparison, y = n_genes, fill = regulation)) +
        ggplot2::geom_bar(stat = "identity") +
        ggplot2::labs(title = "Number of Up/Down-regulated Genes per Comparison", x = "Comparison", y = "Number of Genes") +
        ggplot2::theme_minimal() +
        ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45, hjust = 1))
    ggplot2::ggsave(file.path(cluster_dge_inspect_env$plot_pdf_dir, "stacked_barchart.pdf"), p, width = 10, height = 6)
}

# 3. Volcano plots for each comparison
cluster_dge_inspect_env$plot_volcano_plots <- function(dge_subset, lfc_threshold = 0.25, pval_threshold = 0.05) {
    comparisons <- unique(dge_subset$comparison)
    for (comp in comparisons) {
        comp_df <- dge_subset |> dplyr::filter(comparison == comp)
        comp_df <- comp_df |>
            dplyr::mutate(
                regulation = dplyr::case_when(
                    avg_log2FC >= lfc_threshold & p_val_adj < pval_threshold ~ "Up",
                    avg_log2FC <= -lfc_threshold & p_val_adj < pval_threshold ~ "Down",
                    TRUE ~ "NS"
                )
            )

        # Identify top 10 up- and down-regulated genes
        top_up <- comp_df |>
            dplyr::filter(regulation == "Up") |>
            dplyr::arrange(desc(avg_log2FC)) |>
            dplyr::slice_head(n = 10)

        top_down <- comp_df |>
            dplyr::filter(regulation == "Down") |>
            dplyr::arrange(avg_log2FC) |>
            dplyr::slice_head(n = 10)

        # Create label column
        comp_df <- comp_df |>
            dplyr::mutate(
                label = dplyr::case_when(
                    gene %in% top_up$gene ~ gene,
                    gene %in% top_down$gene ~ gene,
                    TRUE ~ ""
                )
            )

        write.csv(comp_df, file = file.path(cluster_dge_inspect_env$plot_data_dir, paste0("volcano_", comp, ".csv")), row.names = FALSE)
        p <- ggplot(comp_df, aes(x = avg_log2FC, y = -log10(p_val_adj), color = regulation)) +
            ggplot2::geom_point(alpha = 0.6) +
            ggrepel::geom_text_repel(aes(label = label), max.overlaps = 20, size = 3) +
            ggplot2::scale_color_manual(values = c("Up" = "red", "Down" = "blue", "NS" = "grey")) +
            ggplot2::labs(title = paste("Volcano Plot:", comp), x = "avg_log2FC", y = "-log10(p_val_adj)") +
            ggplot2::theme_minimal()
        ggplot2::ggsave(file.path(cluster_dge_inspect_env$plot_pdf_dir, paste0("volcano_", comp, ".pdf")), p, width = 7, height = 6)
    }
}

# Main function to run all plots
cluster_dge_inspect_env$run_all_plots <- function(dge_csv_path) {
    dge <- cluster_dge_inspect_env$load_dge_data(dge_csv_path)
    # Split DGE data by group
    if (!"group" %in% colnames(dge)) {
        stop("Column 'group' not found in DGE data. Please ensure the data contains a 'group' column.")
    }

    groups <- unique(dge$group)
    cat("Found groups:", paste(groups, collapse = ", "), "\n")

    for (group_name in groups) {
        cat("Processing group:", group_name, "\n")
        dge_subset <- dge |> dplyr::filter(group == group_name)

        # Update output directories for this group
        group_plot_data_dir <- file.path(cluster_dge_inspect_env$plot_data_dir, group_name)
        group_plot_pdf_dir <- file.path(cluster_dge_inspect_env$plot_pdf_dir, group_name)
        dir.create(group_plot_data_dir, showWarnings = FALSE, recursive = TRUE)
        dir.create(group_plot_pdf_dir, showWarnings = FALSE, recursive = TRUE)

        # Temporarily update environment paths
        original_data_dir <- cluster_dge_inspect_env$plot_data_dir
        original_pdf_dir <- cluster_dge_inspect_env$plot_pdf_dir
        cluster_dge_inspect_env$plot_data_dir <- group_plot_data_dir
        cluster_dge_inspect_env$plot_pdf_dir <- group_plot_pdf_dir

        # Run plots for this group subset
        cluster_dge_inspect_env$plot_comparison_scatterplots(dge_subset)
        cluster_dge_inspect_env$plot_stacked_barchart(dge_subset)
        cluster_dge_inspect_env$plot_volcano_plots(dge_subset)

        # Restore original paths
        cluster_dge_inspect_env$plot_data_dir <- original_data_dir
        cluster_dge_inspect_env$plot_pdf_dir <- original_pdf_dir
    }
}


# ==========================================
# Run all plots for the DGE on clusters, cell subtypes, and [optionally] combined cell types
# ==========================================

dge_types <- list(
    list(
        csv_path = paste0(output_dirs$deg_clusters, "merged_dge_on_clusters.csv"),
        plot_data_dir = paste0(output_dirs$deg_clusters, "inspection_plot_data/"),
        plot_pdf_dir = paste0(output_dirs$deg_clusters, "inspection_plots_pdf/")
    ),
    list(
        csv_path = paste0(output_dirs$deg_celltypes, "merged_dge_on_cellTypes.csv"),
        plot_data_dir = paste0(output_dirs$deg_celltypes, "inspection_plot_data/"),
        plot_pdf_dir = paste0(output_dirs$deg_celltypes, "inspection_plots_pdf/")
    )
)

# Optionally add combine_celltypes if present
if (!is.null(output_dirs$deg_combine_celltypes)) {
    dge_types[[length(dge_types) + 1]] <- list(
        csv_path = paste0(output_dirs$deg_combine_celltypes, "merged_dge_on_cellTypes.csv"),
        plot_data_dir = paste0(output_dirs$deg_combine_celltypes, "inspection_plot_data/"),
        plot_pdf_dir = paste0(output_dirs$deg_combine_celltypes, "inspection_plots_pdf/")
    )
}

for (dge_info in dge_types) {
    dge_csv_path <- dge_info$csv_path
    if (!file.exists(dge_csv_path)) {
        warning("Merged DGE CSV file not found. Please generate it before running this script: ", dge_csv_path)
        next
    }
    cat("Processing DGE file:", dge_csv_path, "\n")
    cluster_dge_inspect_env$plot_data_dir <- dge_info$plot_data_dir
    cluster_dge_inspect_env$plot_pdf_dir <- dge_info$plot_pdf_dir
    dir.create(cluster_dge_inspect_env$plot_data_dir, showWarnings = FALSE, recursive = TRUE)
    dir.create(cluster_dge_inspect_env$plot_pdf_dir, showWarnings = FALSE, recursive = TRUE)
    cluster_dge_inspect_env$run_all_plots(dge_csv_path)
}

File: DGE/2_DGE_inspection_plots_from_csv.py
"""
===============================================================================
CLUSTER DGE INSPECTION PLOTS SCRIPT (Python Version)
===============================================================================
Purpose:
  Python version of the R script that generates inspection plots for merged 
  differential gene expression (DGE) results on clusters, cell subtypes, and 
  optionally combined cell types.

What are "Inspection Plots"?
  Diagnostic and summary visualizations designed to help users quickly assess,
  compare, and interpret the results of DGE analyses:
    - Scatterplots: Compare log2 fold changes for shared genes between comparisons
    - Stacked Bar Charts: Show number of up- and down-regulated genes per comparison
    - Volcano Plots: Display relationship between fold change and statistical significance

Usage:
  python 2_DGE_inspection_plots_from_csv.py

Dependencies:
  - pandas, numpy, matplotlib, seaborn, itertools, yaml
  - Expects merged DGE CSV files to be present in output directories

Author: Generated from R script
Date: 2024
===============================================================================
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from itertools import combinations
import yaml
import os
import warnings

class DGEInspectionPlots:
    def __init__(self):
        self.load_config()
        # Set ggplot style and better color palette
        plt.style.use('ggplot')
        self.colors = {
            'Up': '#E31A1C',      # Red
            'Down': '#1F78B4',    # Blue  
            'NS': '#CCCCCC',      # Light gray
            'scatter': '#D62728', # Dark red for scatter points
            'line': '#7F7F7F'     # Gray for reference lines
        }
        
    def load_config(self):
        """Load configuration from YAML file"""
        with open('config/batch_config.yaml', 'r') as f:
            self.config = yaml.safe_load(f)
        self.output_dirs = self.config['output_dirs']
        
    def standardize_dge_columns(self, dge):
        """Standardize column names for different DGE file formats"""
        # Check if this is the cellTypes format
        if 'names' in dge.columns and 'logfoldchanges' in dge.columns:
            # Map cellTypes format to standard format
            column_mapping = {
                'names': 'gene',
                'logfoldchanges': 'avg_log2FC',
                'pvals_adj': 'p_val_adj',
                'pvals': 'p_val'
            }
            dge = dge.rename(columns=column_mapping)
            
            # Add missing columns with default values if they don't exist
            if 'pct.1' not in dge.columns:
                dge['pct.1'] = np.nan
            if 'pct.2' not in dge.columns:
                dge['pct.2'] = np.nan
                
            # Rename cell_type to region for consistency (if region doesn't exist)
            if 'cell_type' in dge.columns and 'region' not in dge.columns:
                dge = dge.rename(columns={'cell_type': 'region'})
        
        # Ensure avg_log2FC is numeric
        dge['avg_log2FC'] = pd.to_numeric(dge['avg_log2FC'], errors='coerce')
        return dge
    
    def load_dge_data(self, dge_csv_path):
        """Load merged DGE data"""
        dge = pd.read_csv(dge_csv_path)
        # Standardize column names based on format
        dge = self.standardize_dge_columns(dge)
        return dge
    
    def plot_comparison_scatterplots(self, dge_subset, plot_data_dir, plot_pdf_dir, 
                                   lfc_threshold=0.25, pval_threshold=0.05):
        """Generate scatterplots for all pairs of comparisons (shared genes)"""
        comparisons = dge_subset['comparison'].unique()
        
        for comp1, comp2 in combinations(comparisons, 2):
            d1 = dge_subset[(dge_subset['comparison'] == comp1) & 
                           (dge_subset['p_val_adj'] < pval_threshold)]
            d2 = dge_subset[(dge_subset['comparison'] == comp2) & 
                           (dge_subset['p_val_adj'] < pval_threshold)]
            
            shared_genes = set(d1['gene']).intersection(set(d2['gene']))
            if len(shared_genes) == 0:
                continue
                
            d1_shared = d1[d1['gene'].isin(shared_genes)][['gene', 'avg_log2FC']].drop_duplicates('gene')
            d2_shared = d2[d2['gene'].isin(shared_genes)][['gene', 'avg_log2FC']].drop_duplicates('gene')
            
            plot_df = pd.merge(d1_shared, d2_shared, on='gene', suffixes=('_1', '_2'))
            plot_df['size'] = abs(plot_df['avg_log2FC_1']) + abs(plot_df['avg_log2FC_2'])
            
            # Save data
            plot_df.to_csv(os.path.join(plot_data_dir, f'scatter_{comp1}_against_{comp2}.csv'), 
                          index=False)
            
            # Create plot with ggplot styling
            fig, ax = plt.subplots(figsize=(8, 8))
            
            # Add reference lines
            ax.axhline(y=0, linestyle='--', color=self.colors['line'], alpha=0.6, linewidth=1)
            ax.axvline(x=0, linestyle='--', color=self.colors['line'], alpha=0.6, linewidth=1)
            
            # Create scatter plot with size mapping
            scatter = ax.scatter(plot_df['avg_log2FC_1'], plot_df['avg_log2FC_2'], 
                               s=plot_df['size']*15, alpha=0.7, color=self.colors['scatter'],
                               edgecolors='white', linewidth=0.5)
            
            # Add labels for most extreme points
            extreme_points = plot_df.nlargest(10, 'size')
            for _, row in extreme_points.iterrows():
                ax.annotate(row['gene'], (row['avg_log2FC_1'], row['avg_log2FC_2']), 
                           xytext=(8, 8), textcoords='offset points', fontsize=9,
                           bbox=dict(boxstyle='round,pad=0.3', facecolor='white', alpha=0.8),
                           arrowprops=dict(arrowstyle='->', connectionstyle='arc3,rad=0'))
            
            ax.set_xlabel(f'avg_log2FC ({comp1})', fontsize=12, fontweight='bold')
            ax.set_ylabel(f'avg_log2FC ({comp2})', fontsize=12, fontweight='bold')
            ax.set_title(f'Shared DE Genes: {comp1} vs {comp2}', fontsize=14, fontweight='bold', pad=20)
            
            # Improve grid and spines
            ax.grid(True, alpha=0.3)
            ax.spines['top'].set_visible(False)
            ax.spines['right'].set_visible(False)
            
            plt.tight_layout()
            plt.savefig(os.path.join(plot_pdf_dir, f'scatter_{comp1}_against_{comp2}.pdf'), 
                       dpi=300, bbox_inches='tight')
            plt.close()
    
    def plot_stacked_barchart(self, dge_subset, plot_data_dir, plot_pdf_dir,
                             lfc_threshold=0.25, pval_threshold=0.05):
        """Generate stacked bar chart: number of up/down-regulated genes per comparison"""
        def classify_regulation(row):
            if row['avg_log2FC'] >= lfc_threshold and row['p_val_adj'] < pval_threshold:
                return 'Up'
            elif row['avg_log2FC'] <= -lfc_threshold and row['p_val_adj'] < pval_threshold:
                return 'Down'
            else:
                return 'NS'
        
        # Make a copy to avoid SettingWithCopyWarning
        dge_subset = dge_subset.copy()
        dge_subset.loc[:, 'regulation'] = dge_subset.apply(classify_regulation, axis=1)
        
        bar_df = (dge_subset[dge_subset['regulation'] != 'NS']
                 .groupby(['comparison', 'regulation'])
                 .size()
                 .reset_index(name='n_genes'))
        
        # Save data
        bar_df.to_csv(os.path.join(plot_data_dir, 'stacked_barchart_data.csv'), index=False)
        
        # Create plot with ggplot styling
        pivot_df = bar_df.pivot(index='comparison', columns='regulation', values='n_genes').fillna(0)
        
        fig, ax = plt.subplots(figsize=(12, 7))
        
        # Create stacked bar chart with better colors
        colors_list = [self.colors['Down'], self.colors['Up']]
        pivot_df.plot(kind='bar', stacked=True, color=colors_list, ax=ax, 
                     edgecolor='white', linewidth=0.5)
        
        ax.set_title('Number of Up/Down-regulated Genes per Comparison', 
                    fontsize=14, fontweight='bold', pad=20)
        ax.set_xlabel('Comparison', fontsize=12, fontweight='bold')
        ax.set_ylabel('Number of Genes', fontsize=12, fontweight='bold')
        
        # Improve legend
        ax.legend(title='Regulation', title_fontsize=11, fontsize=10, 
                 loc='upper right', frameon=True, fancybox=True, shadow=True)
        
        # Rotate x-axis labels
        plt.setp(ax.get_xticklabels(), rotation=45, ha='right')
        
        # Improve grid and spines
        ax.grid(True, alpha=0.3, axis='y')
        ax.spines['top'].set_visible(False)
        ax.spines['right'].set_visible(False)
        
        plt.tight_layout()
        plt.savefig(os.path.join(plot_pdf_dir, 'stacked_barchart.pdf'), 
                   dpi=300, bbox_inches='tight')
        plt.close()
    
    def plot_volcano_plots(self, dge_subset, plot_data_dir, plot_pdf_dir,
                          lfc_threshold=0.25, pval_threshold=0.05):
        """Generate volcano plots for each comparison"""
        def classify_regulation(row):
            if row['avg_log2FC'] >= lfc_threshold and row['p_val_adj'] < pval_threshold:
                return 'Up'
            elif row['avg_log2FC'] <= -lfc_threshold and row['p_val_adj'] < pval_threshold:
                return 'Down'
            else:
                return 'NS'
        
        comparisons = dge_subset['comparison'].unique()
        
        for comp in comparisons:
            comp_df = dge_subset[dge_subset['comparison'] == comp].copy()
            comp_df['regulation'] = comp_df.apply(classify_regulation, axis=1)
            
            # Fix: Replace zero or negative p_val_adj with small positive value to avoid log10 errors
            comp_df['p_val_adj'] = comp_df['p_val_adj'].clip(lower=1e-300)
            
            # Identify top 10 up- and down-regulated genes
            top_up = (comp_df[comp_df['regulation'] == 'Up']
                     .nlargest(10, 'avg_log2FC'))
            top_down = (comp_df[comp_df['regulation'] == 'Down']
                       .nsmallest(10, 'avg_log2FC'))
            
            # Create label column
            comp_df['label'] = ''
            comp_df.loc[comp_df['gene'].isin(top_up['gene']), 'label'] = comp_df['gene']
            comp_df.loc[comp_df['gene'].isin(top_down['gene']), 'label'] = comp_df['gene']
            
            # Save data
            comp_df.to_csv(os.path.join(plot_data_dir, f'volcano_{comp}.csv'), index=False)
            
            # Create plot with ggplot styling
            fig, ax = plt.subplots(figsize=(9, 7))
            
            # Plot points by regulation type with better styling
            for reg_type in ['NS', 'Down', 'Up']:  # Plot NS first so colored points are on top
                subset = comp_df[comp_df['regulation'] == reg_type]
                if len(subset) > 0:
                    alpha = 0.4 if reg_type == 'NS' else 0.7
                    size = 20 if reg_type != 'NS' else 15
                    ax.scatter(subset['avg_log2FC'], -np.log10(subset['p_val_adj']), 
                             color=self.colors[reg_type], alpha=alpha, s=size,
                             label=f'{reg_type} ({len(subset)})', 
                             edgecolors='white' if reg_type != 'NS' else 'none', 
                             linewidth=0.3)
            
            # Add significance and fold change threshold lines
            if pval_threshold > 0:
                ax.axhline(y=-np.log10(pval_threshold), linestyle='--', 
                          color=self.colors['line'], alpha=0.6, linewidth=1)
            ax.axvline(x=lfc_threshold, linestyle='--', 
                      color=self.colors['line'], alpha=0.6, linewidth=1)
            ax.axvline(x=-lfc_threshold, linestyle='--', 
                      color=self.colors['line'], alpha=0.6, linewidth=1)
            
            # Add labels for top genes
            labeled_genes = comp_df[comp_df['label'] != '']
            for _, row in labeled_genes.iterrows():
                ax.annotate(row['label'], (row['avg_log2FC'], -np.log10(row['p_val_adj'])), 
                           xytext=(8, 8), textcoords='offset points', fontsize=9,
                           bbox=dict(boxstyle='round,pad=0.3', facecolor='white', alpha=0.8),
                           arrowprops=dict(arrowstyle='->', connectionstyle='arc3,rad=0.1'))
            
            ax.set_xlabel('avg_log2FC', fontsize=12, fontweight='bold')
            ax.set_ylabel('-log10(p_val_adj)', fontsize=12, fontweight='bold')
            ax.set_title(f'Volcano Plot: {comp}', fontsize=14, fontweight='bold', pad=20)
            
            # Improve legend
            ax.legend(title='Regulation', title_fontsize=11, fontsize=10, 
                     loc='upper right', frameon=True, fancybox=True, shadow=True)
            
            # Improve grid and spines
            ax.grid(True, alpha=0.3)
            ax.spines['top'].set_visible(False)
            ax.spines['right'].set_visible(False)
            
            plt.tight_layout()
            plt.savefig(os.path.join(plot_pdf_dir, f'volcano_{comp}.pdf'), 
                       dpi=300, bbox_inches='tight')
            plt.close()
    
    def run_all_plots(self, dge_csv_path, plot_data_dir, plot_pdf_dir):
        """Main function to run all plots"""
        dge = self.load_dge_data(dge_csv_path)
        
        # Check for group column
        if 'group' not in dge.columns:
            raise ValueError("Column 'group' not found in DGE data. Please ensure the data contains a 'group' column.")
        
        groups = dge['group'].unique()
        print(f"Found groups: {', '.join(groups)}")
        
        for group_name in groups:
            print(f"Processing group: {group_name}")
            dge_subset = dge[dge['group'] == group_name]
            
            # Create group-specific directories
            group_plot_data_dir = os.path.join(plot_data_dir, group_name)
            group_plot_pdf_dir = os.path.join(plot_pdf_dir, group_name)
            os.makedirs(group_plot_data_dir, exist_ok=True)
            os.makedirs(group_plot_pdf_dir, exist_ok=True)
            
            # Run plots for this group subset
            self.plot_comparison_scatterplots(dge_subset, group_plot_data_dir, group_plot_pdf_dir)
            self.plot_stacked_barchart(dge_subset, group_plot_data_dir, group_plot_pdf_dir)
            self.plot_volcano_plots(dge_subset, group_plot_data_dir, group_plot_pdf_dir)

def main():
    """Main execution function"""
    plotter = DGEInspectionPlots()
    
    # Define DGE types to process
    dge_types = [
        {
            'csv_path': os.path.join(plotter.output_dirs['deg_clusters'], 'merged_dge_on_clusters.csv'),
            'plot_data_dir': os.path.join(plotter.output_dirs['deg_clusters'], 'inspection_plot_data/'),
            'plot_pdf_dir': os.path.join(plotter.output_dirs['deg_clusters'], 'inspection_plots_pdf/')
        },
        {
            'csv_path': os.path.join(plotter.output_dirs['deg_celltypes'], 'merged_dge_on_cellTypes.csv'),
            'plot_data_dir': os.path.join(plotter.output_dirs['deg_celltypes'], 'inspection_plot_data/'),
            'plot_pdf_dir': os.path.join(plotter.output_dirs['deg_celltypes'], 'inspection_plots_pdf/')
        }
    ]
    
    # Optionally add combine_celltypes if present
    if 'deg_combine_celltypes' in plotter.output_dirs:
        dge_types.append({
            'csv_path': os.path.join(plotter.output_dirs['deg_combine_celltypes'], 'merged_dge_on_cellTypes.csv'),
            'plot_data_dir': os.path.join(plotter.output_dirs['deg_combine_celltypes'], 'inspection_plot_data/'),
            'plot_pdf_dir': os.path.join(plotter.output_dirs['deg_combine_celltypes'], 'inspection_plots_pdf/')
        })
    
    # Process each DGE type
    for dge_info in dge_types:
        dge_csv_path = dge_info['csv_path']
        if not os.path.exists(dge_csv_path):
            warnings.warn(f"Merged DGE CSV file not found. Please generate it before running this script: {dge_csv_path}")
            continue
            
        print(f"Processing DGE file: {dge_csv_path}")
        
        # Create directories
        os.makedirs(dge_info['plot_data_dir'], exist_ok=True)
        os.makedirs(dge_info['plot_pdf_dir'], exist_ok=True)
        
        # Run all plots
        plotter.run_all_plots(dge_csv_path, dge_info['plot_data_dir'], dge_info['plot_pdf_dir'])

if __name__ == "__main__":
    main()
